lib/config/Configuration.js

```diff
@@ -0,0 +1,332 @@
+
+/**
+ * @public
+ * @typedef @ui5/project/config/Configuration~cacheMode
+ * @property {string} "default" Cache everything, invalidate after 9 hours
+ * @property {string} "off" Do not cache
+ * @property {string} "force" Use cache only - no requests
+ * @property {string} "relaxed" Try to refresh, but fallback to cache if refresh fails for up to one week
+ */
+
+/**
+ *
+ *
+ * @public
+ * @class
+ * @alias @ui5/project/config/Configuration
+ */
+export class Configuration {
+	#cwd;
+	#sources;
+	#version;
+	#cacheMode;
+	#ui5HomeDir;
+	#snapshotEndpointUrl;
+	#frameworkDir;
+	#artifactsDir;
+	#packagesDir;
+	#metadataDir;
+	#stagingDir;
+	#lockDir;
+	#cacheDir;
+
+	/**
+	 * @param {object} configuration
+	 * @param {string} [configuration.cwd="./"]
+	 * @param {string} [configuration.version=""]
+	 * @param {string} [configuration.sources=false]
+	 * @param {string} [configuration.ui5HomeDir="~/.ui5/"]
+	 * @param {string} [configuration.snapshotEndpointUrl=""]
+	 * @param {@ui5/project/config/Configuration~cacheMode} [configuration.cacheMode="default"]
+	 * @param {string} [configuration.frameworkDir="~/${ui5HomeDir}/framework"]
+	 *   Directory to store framework artifacts and metadata in.
+	 * @param {string} [configuration.cacheDir="~/${ui5HomeDir}/cacache"]
+	 * @param {string} [configuration.artifactsDir="~/${ui5HomeDir}/artifacts"]
+	 * @param {string} [configuration.packagesDir="~/${ui5HomeDir}/packages"]
+	 * @param {string} [configuration.metadataDir="~/${ui5HomeDir}/metadata"]
+	 * @param {string} [configuration.stagingDir="~/${ui5HomeDir}/staging"]
+	 * @param {string} [configuration.lockDir="~/${ui5HomeDir}/locks"]
+	 */
+	constructor({
+		cwd,
+		version,
+		sources,
+		ui5HomeDir,
+		snapshotEndpointUrl,
+		cacheMode = "default",
+		frameworkDir,
+		cacheDir,
+		artifactsDir,
+		packagesDir,
+		metadataDir,
+		stagingDir,
+		lockDir
+	}) {
+		this.#cwd = cwd ? path.resolve(cwd) : process.cwd();
+		this.#sources = !!sources;
+		this.#version = version;
+		this.#cacheMode = cacheMode;
+		this.#ui5HomeDir = ui5HomeDir ? path.resolve(ui5HomeDir) : path.join(os.homedir(), ".ui5");
+
+		this.#snapshotEndpointUrl = process.env.UI5_MAVEN_SNAPSHOT_ENDPOINT || snapshotEndpointUrl;
+		this.#frameworkDir = frameworkDir ? path.resolve(frameworkDir) : path.join(this.#ui5HomeDir, "framework");
+		this.#artifactsDir = artifactsDir ?
+			path.resolve(artifactsDir) : path.join(this.#frameworkDir, "artifacts");
+		this.#packagesDir = packagesDir ?
+			path.resolve(packagesDir) : path.join(this.#frameworkDir, "packages");
+		this.#metadataDir = metadataDir ?
+			path.resolve(metadataDir) : path.join(this.#frameworkDir, "metadata");
+		this.#stagingDir = stagingDir ? path.resolve(stagingDir) : path.join(this.#frameworkDir, "staging");
+		this.#lockDir = lockDir ? path.resolve(lockDir) : path.join(this.#frameworkDir, "locks");
+		this.#cacheDir = cacheDir ? path.resolve(cacheDir) : path.join(this.#frameworkDir, "cacache");
```

Comment:
There are several questions that arise here.

### Which properties/settings should be within the Configuration object?

Currently, all the needed configuration options are handled here. The reasoning is that many depend on each other i.e. `frameworkDir` depends on `ui5HomeDir`

On the other side, the core reasoning to introduce such a Configuration was to support the Snapshot URL.

### Do we need default values?

According to the (spec)[https://github.com/SAP/ui5-tooling/blob/rfc-configuration/rfcs/0013-configuration.md], there shouldn't be default values.

Properties could be defined/provided separately. But the way we use them internally implies that they depend on each other i.e. `frameworkDir` depends on `ui5HomeDir`

Reply:
For this iteration, we have decided to go only with the necessary `snapshotEndpointUrl`. The questions above would be resolved in separate iteration(s)/BLI(s).